src/VisualStudio/Core/Def/Implementation/FindReferences/AsyncFindReferencesPresenter.DataSource.cs

```diff
@@ -0,0 +1,180 @@
+            {
+                _asyncListener = asyncListener;
+            }
+
+            public override CancellationToken CancellationToken => _cancellationTokenSource.Token;
+
+            #region ITableDataSource
+
+            public string DisplayName => "Roslyn Data Source";
+
+            public string Identifier => "Roslyn Identifier";
+
+            public string SourceTypeIdentifier => "FindAllReferencesProvider";
+
+            public int CurrentVersionNumber { get; private set; }
+
+            public IDisposable Subscribe(ITableDataSink sink)
+            {
+                var subscription = new Subscription(_cancellationTokenSource, sink);
+                _subscriptions.Add(subscription);
+
+                sink.AddFactory(this, removeAllFactories: true);
+
+                return subscription;
+            }
+
+            #endregion
+
+            #region EditorFindReferencesContext overrides.
+
+            public override void OnStarted()
+            {
+                foreach (var subscription in _subscriptions)
+                {
+                    subscription.TableDataSink.IsStable = false;
+                }
+            }
+
+            public override void OnCompleted()
+            {
+                foreach (var subscription in _subscriptions)
+                {
+                    subscription.TableDataSink.IsStable = true;
+                }
+            }
+
+            public async override void OnReferenceFound(INavigableItem definition, INavigableItem reference)
+            {
+                var workspace = reference.Document.Project.Solution.Workspace as VisualStudioWorkspaceImpl;
+                if (workspace == null)
+                {
+                    return;
+                }
+
+                var projectGuid = workspace.GetHostProject(reference.Document.Project.Id)?.Guid;
+                if (projectGuid == null)
+                {
+                    return;
+                }
+                
+                try
+                {
+                    // We're told about this reference synchronously, but we need to get the 
+                    // SourceText for the reference's Document so that we can determine things
+                    // like it's line/column/text.  We don't want to block this method getting
+                    // that data, so instead we just fire off the async work to get the text
+                    // and use it.  Because we're starting some async work, let the test harness
+                    // know so that it doesn't verify results until this completes.
+                    using (var token = _asyncListener.BeginAsyncOperation(nameof(OnReferenceFound)))
+                    {
+                        await OnReferenceFoundAsync(definition, reference, projectGuid.Value).ConfigureAwait(false);
+                    }
+                }
+                catch (Exception e) when (FatalError.ReportWithoutCrashUnlessCanceled(e))
```

Comment:
> Do we expect this to be throwing exceptions...?

Cancellation exceptions?  Definitely.   Other exceptions?  No...   But note that i'm kicking off async work in a sync-callback.  If i don't do this try/catch, then an actual faulting exception would be unreported.  This allows us to make sure that we actually get the watson.